<?xml version="1.0"?>
<launch>
    <arg name="PROJECT_NAME" value="nonholo_control"/>
	<arg name="avoid_duplicates" value="true"/>
	<param name="xy_tolerance" type="double" value="0.5"/>
	<rosparam command="load" file="$(find nonholo_control)/param/nonholo_params.yaml" />
    <!--include file="$(find final_assignment)/launch/simulation_gmapping.launch" /-->

    <!--node name="$(arg PROJECT_NAME)_redirect_server" pkg="nonholo_control" type="$(arg PROJECT_NAME)_redirect_server" /-->
    <node name="$(arg PROJECT_NAME)_gotopo_wllflw_redirect" pkg="nonholo_control" type="$(arg PROJECT_NAME)_gotopo_wllflw_redirect" />

    <node name="mux_cmdvel" pkg="topic_tools" type="mux" args="cmd_vel cmd_vel_move cmd_vel_bug cmd_vel_stop mux:=mux_cmdvel"/>
    
    <!--node name="mux_wllflw_nhc" pkg="topic_tools" type="mux" args="wall_follower_switch_nhc wall_follower_switch redirect   mux:=mux_wllflw_nhc"/-->
    <!--node name="mux_wllflw_bug" pkg="topic_tools" type="mux" args="wall_follower_switch_bug redirect wall_follower_switch   mux:=mux_wllflw_bug"/-->
    <!--node name="mux_gotopo_nhc" pkg="topic_tools" type="mux" args="go_to_point_switch_nhc   go_to_point_switch redirect       mux:=mux_gotopo_nhc"/-->
    <!--node name="mux_gotopo_bug" pkg="topic_tools" type="mux" args="go_to_point_switch_bug   redirect go_to_point_switch       mux:=mux_gotopo_bug"/-->
    
    
    <!--node name="mux_wllflw" pkg="topic_tools" type="mux" args="wall_follow_switch wllflw_nhc wllflw_bug mux:=mux_wllflw"/-->
    <!--node name="mux_gotopo" pkg="topic_tools" type="mux" args="go_to_point_switch gotopo_nhc gotopo_bug mux:=mux_gotopo"/-->

	<!-- Nodes and parameters performing the bug0 (and wall_follow) algorithm -->
    <arg name="des_x" default= "0"/>
    <arg name="des_y" default= "0"/>
    <param name="des_pos_x" value="$(arg des_x)" />
    <param name="des_pos_y" value="$(arg des_y)" />
    <node pkg="final_assignment" type="wall_follow_service_m.py" name="wall_follower">
        <remap from="cmd_vel" to="cmd_vel_bug"/>
    </node>
    <node pkg="final_assignment" type="go_to_point_service_m.py" name="go_to_point">
        <remap from="cmd_vel" to="cmd_vel_bug"/>
    </node>
    <!--node pkg="robot_description" type="user_interface.py" name="user_interface" output="screen" /-->
    <node pkg="final_assignment" type="bug_m.py" name="bug" respawn="true">
        <remap from="user_interface" to="redirect_bug_user_interface"/>
        <remap from="wall_follower_switch" to="wall_follower_switch_bug"/>
        <remap from="go_to_point_switch" to="go_to_point_switch_bug"/>
    </node>
	
	<!-- Nodes and parameters performing the move_base algorithm -->
	<node pkg="move_base" type="move_base" respawn="true" name="move_base">
        <rosparam file="$(find final_assignment)/param/costmap_common_params.yaml" command="load" ns="global_costmap"/>
        <rosparam file="$(find final_assignment)/param/costmap_common_params.yaml" command="load" ns="local_costmap"/>
        <rosparam file="$(find final_assignment)/param/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find final_assignment)/param/global_costmap_params.yaml" command="load"/>
        <rosparam file="$(find final_assignment)/param/base_local_planner_params.yaml" command="load"/>
        <rosparam file="$(find final_assignment)/param/move_base_params.yaml" command="load"/>
        <remap from="cmd_vel" to="cmd_vel_move"/>
    </node>
  
   <!-- The nodes of the overall control part, implemented in this package -->
    
    <node name="$(arg PROJECT_NAME)_unreachable_target_detection" pkg="nonholo_control" type="$(arg PROJECT_NAME)_unreachable_target_detection" output="screen"/>
    <node name="$(arg PROJECT_NAME)_target_reached_detection" pkg="nonholo_control" type="$(arg PROJECT_NAME)_target_reached_detection" output="screen" />
   
    <node name="$(arg PROJECT_NAME)_random_position_server" pkg="nonholo_control" type="$(arg PROJECT_NAME)_random_position_server" 
	      args="$(arg avoid_duplicates)"/>
	<node name="$(arg PROJECT_NAME)_user_position_server" pkg="nonholo_control" type="$(arg PROJECT_NAME)_user_position_server" output="screen"/>
	<node name="$(arg PROJECT_NAME)_user_interface" pkg="nonholo_control" type="$(arg PROJECT_NAME)_user_interface" required="true" output="screen">
        <remap from="wall_follower_switch" to="wall_follower_switch_nhc"/>
        <remap from="go_to_point_switch" to="go_to_point_switch_nhc"/>
	</node>
	
</launch>
